### PR TITLE
[REST] Add memory pool API

### PR DESCRIPTION
**Pull request by paveljanik**
As requested in #5844, memory pool API added.

Get TX mempool info:

``` shell
$ curl localhost:8332/rest/mempool/info.json 2>/dev/null | json_pp
{
   "bytes" : 877,
   "size" : 2
}
$
```

Get TX mempool contents:

``` shell
$ curl localhost:8332/rest/mempool/contents.json 2>/dev/null | json_pp
{
   "d84fb825b54442d737f9026ac52c4208f48eae8e1aaf9c942acede7ee8457f35" : {
      "currentpriority" : 8060141.48717949,
      "time" : 1429038596,
      "fee" : 0.0001,
      "height" : 352141,
      "depends" : [],
      "startingpriority" : 8060141.48717949,
      "size" : 225
   },
...
$ 
```

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK

**Comment by jgarzik** [comment]
ACK

**Comment by jonasschnelli** [comment]
Concept ACK.
Misses RPC test.

**Comment by paveljanik** [comment]
Yup, I plan to add them in the evening or tomorrow,

File: src/rest.cpp [hunk 1]
```diff
@@ -252,6 +252,60 @@ static bool rest_chaininfo(AcceptedConnection* conn,
...
+{
+    vector<string> params;
+    const RetFormat rf = ParseDataFormat(params, strReq);
+
+    switch (rf) {
+    case RF_JSON: {
+        Array rpcParams;
+	rpcParams.push_back(true);
+
+        Value mempoolObject = getrawmempool(rpcParams, false);
```
Comment (jonasschnelli) [inline src/rest.cpp]: I'm not sure if we should just loop through RPC commands. I know that i did the same when i added the `/chaininfos` REST command. This will tie the REST interface to the current RPC implementation. But duplicating code is also something we should avoid.
I don't have a solution here i just wanted to bring up this concern.

File: src/rest.cpp [hunk 2]
```diff
@@ -252,6 +252,60 @@ static bool rest_chaininfo(AcceptedConnection* conn,
...
+                                   const std::map<std::string, std::string>& mapHeaders,
+                                   bool fRun)
+{
+    vector<string> params;
+    const RetFormat rf = ParseDataFormat(params, strReq);
+
+    switch (rf) {
+    case RF_JSON: {
+        Array rpcParams;
+	rpcParams.push_back(true);
```
Comment (jonasschnelli) [inline src/rest.cpp]: very little nit: indent.

**Comment by paveljanik** [comment]
nit addressed, squashed, ready.

**Comment by jonasschnelli** [comment]
Tested ACK.

**Comment by paveljanik** [comment]
Other reviews, please?

**Comment by jgarzik** [comment]
tested ACK

**Comment by jonasschnelli** [comment]
Needs rebase.

**Comment by paveljanik** [comment]
Rebased (changed the header style in README to match the other entries).

**Comment by paveljanik** [comment]
Needs to be adapted to UniValue. Tomorrow.

**Comment by paveljanik** [comment]
Travis CI build fails only in one environment because of the comparison tool.
Ready for merge.

**Comment by jonasschnelli** [comment]
I like this PR.

But, am i alone with the feeling that calling a RPC method (`getrawmempool`, `getmempoolinfo`) will end up in REST/RPC dependency and smells after a unideal design? Duplicating code would also be kinda stupid. Maybe factoring out some code to a independent helper class/file?

But however reACK.

**Comment by paveljanik** [comment]
@jonasschnelli Something like CTxMemPool::toJSON factored out from get*mempool*? Or do you prefer interface similar to TxToJSON (defined in rpc, used in rest)?

I do not have a problem to refactor once agreed, but in the next PR please.

**Comment by sipa** [comment]
No JSON conversion code inside core classes, please. That's something that belongs in the RPC glue layer.

**Comment by jonasschnelli** [comment]
For sure we don't want json code within core classes. But i think calling a RPC function within REST should be avoided. Maybe something like `blockToJSON` or similar. Just factor out the json encoding and  for now – place it within the `rpcblockchain.cpp`.

**Comment by paveljanik** [comment]
@jonasschnelli Please review. Thanks.

**Comment by jonasschnelli** [comment]
design wise this looks much better.
utACK. Will give it a real testshot soon.

**Comment by jonasschnelli** [comment]
Tested ACK.
Wrote a simple rpc test for `rawtransaction.py` https://github.com/jonasschnelli/bitcoin/commit/5aa299ff2c40b4b5504ab99ecf19887842d7cf39

**Comment by jonasschnelli** [comment]
Tested ACK.

Two nits:
a) missing support for HEX/BIN (would make it consistence).
aa) could be serialized vector of transactions for `mempool/content` and two serialized int64_t for `mempool/info`
b) your diff looks somehow strange (look at `UniValue getrawmempool(const UniValue& params, bool fHelp)` remove L150-185, re-add at L169-232). Maybe shuffle some lines to avoid this?

**Comment by paveljanik** [comment]
@jonasschnelli Thanks for review.

- hex/bin: Does this make sense at all here? I do not think so. I modelled info it similar to `getchaininfo`.
- does git diff --diff-algorithm=minimal help you?

**Comment by jonasschnelli** [comment]
> hex/bin: Does this make sense at all here? I do not think so. I modelled info it similar to getchaininfo.

If someone actively polls the mempool over REST, he could probably save CPU time and bandwidth if there would be a BIN/HEX support. But as said: low prio.

> does git diff --diff-algorithm=minimal help you?

Thanks. Yes: https://github.com/bitcoin/bitcoin/pull/6013/files?1 (the 1 at the end helps)

**Comment by paveljanik** [comment]
Hmm, 1 at the end seems to not have any effect here (compared with two curl calls and diff on the output).

**Comment by paveljanik** [comment]
Rebased.
Added "usage" to the documentation.

**Comment by jgarzik** [comment]
re-ACK

**Comment by jonasschnelli** [comment]
Needs rebase.
Anything holding this back?

**Comment by paveljanik** [comment]
Rebased (to accomodate #6504 changes).
Ready for merge IMO.

**Comment by btcdrak** [comment]
ACK

File: doc/REST-interface.md [hunk 3]
```diff
@@ -77,6 +77,20 @@ $ curl localhost:18332/rest/getutxos/checkmempool/b2cdfd7b89def827ff8af7cd9bff76
 }
 ```
 
+####Memory pool
+`GET /rest/mempool/info.json`
+
+Returns various informations about the TX mempool.
```
Comment (TheBlueMatt) [inline doc/REST-interface.md]: Nit: informations->information

File: doc/REST-interface.md [hunk 4]
```diff
@@ -77,6 +77,20 @@ $ curl localhost:18332/rest/getutxos/checkmempool/b2cdfd7b89def827ff8af7cd9bff76
...
+Only supports JSON as output format.
+* size : (numeric) the number of transactions in the TX mempool
+* bytes : (numeric) size of the TX mempool in bytes
+* usage : (numeric) total TX mempool memory usage
+
+`GET /rest/mempool/contents.json`
+
+Returns transactions in the TX mempool.
+Only supports JSON as output format.
```
Comment (TheBlueMatt) [inline doc/REST-interface.md]: Its not clear you mean that both info and contents only support JSON.

File: src/rest.cpp [hunk 5]
```diff
@@ -65,6 +65,8 @@ class RestErr
 
 extern void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry);
 extern UniValue blockToJSON(const CBlock& block, const CBlockIndex* blockindex, bool txDetails = false);
+extern UniValue mempoolInfoToJSON();
+extern UniValue mempoolToJSON(bool fVerbose = false);
```
Comment (TheBlueMatt) [inline src/rest.cpp]: Nit: rest_chaininfo just aclls the getblockchaininfo RPC function directly. Though your change makes the code a bit cleaner, maybe split the commits up to just call the RPC first, then move the code around in a separate commit.

File: doc/REST-interface.md (hunk 4 above)
Comment (paveljanik) [inline doc/REST-interface.md]: This is why I wrote the same sentence in both parts...

**Comment by TheBlueMatt** [comment]
utACK

File: doc/REST-interface.md (hunk 3 above)
Comment (paveljanik) [inline doc/REST-interface.md]: Fixed in the SQUASHME commit.

File: doc/REST-interface.md (hunk 4 above)
Comment (TheBlueMatt) [inline doc/REST-interface.md]: Oh, missed that :)

File: src/rest.cpp (hunk 5 above)
Comment (paveljanik) [inline src/rest.cpp]: If I should split it into more commits, I'd prefer the other way round:

- separate `mempoolInfoToJSON` and `mempoolToJSON` and call the from RPC functions (ie. prepare the current code for REST)
- add REST functions

Should I really split it?

Comment (TheBlueMatt) [inline src/rest.cpp]: Order them either way. I dont care too much, I just always prefer easier-to-read diffs over fewer commits, but its really up to you. I ACKed it as-is, was more of a suggestion than anything.

**Comment by paveljanik** [comment]
@laanwj Can you please have a look. Thank you.

**Comment by jgarzik** [comment]
I can merge, after this latest commit is squashed.

**Comment by paveljanik** [comment]
Squashed.

File: src/rest.cpp [hunk 6]
```diff
@@ -293,6 +295,58 @@ static bool rest_chaininfo(AcceptedConnection* conn,
     return true; // continue to process further HTTP reqs on this cxn
 }
 
+static bool rest_mempool_info(AcceptedConnection* conn,
+                              const std::string& strURIPart,
+                              const std::string& strRequest,
+                              const std::map<std::string, std::string>& mapHeaders,
+                              bool fRun)
+{
+    vector<string> params;
```
Comment (Diapolo) [inline src/rest.cpp]: You could (in the future) start adding std:: in front of such code, as that means we can more easily get rod of ``using namespace std;`` :).

File: src/rpcblockchain.cpp [hunk 7]
```diff
@@ -757,6 +761,16 @@ UniValue getchaintips(const UniValue& params, bool fHelp)
     return res;
 }
 
+UniValue mempoolInfoToJSON()
+{
+    UniValue ret(UniValue::VOBJ);
+    ret.push_back(Pair("size", (int64_t) mempool.size()));
```
Comment (Diapolo) [inline src/rpcblockchain.cpp]: Minor nit: You also could remove the space before the cast here.

**Comment by jonasschnelli** [comment]
Re ACK

File: src/rpcblockchain.cpp (hunk 7 above)
Comment (paveljanik) [inline src/rpcblockchain.cpp]: Eh? Please read the whole diff (I did not change that code at all).

Space before the cast? I hope we do not have a rule do not add a space after comma...

Comment (Diapolo) [inline src/rpcblockchain.cpp]: Don't be aggressive, I wrote minor nit and I'm well aware you just copied the code. But if there was something left to do you could've just change to ``(int64_t)mempool.size()``, what's the problem. No there is no such rule!

Comment (paveljanik) [inline src/rpcblockchain.cpp]: So you mean *after* the cast? I'd write it the same way it is written anyway.

**Comment by jonasschnelli** [comment]
@Diapolo: i think it's unfair to start nitpicking at this stage of the PR.

**Let's get this PR merged.**

Anyone can open a cleanup PR over @trivial.

**Comment by Diapolo** [comment]
ACK and merge is fine with me, I just comment when I've got time or start reading a pull, nothing more ;).

**Comment by dcousens** [comment]
utACK

File: src/rest.cpp (hunk 2 above)
Comment (KaycoinsInc) [reply src/rest.cpp]: {curl http://localhost:8332/rest/mempool/info.json}
``
